### PR TITLE
feat: update to swift 5 and refactor to use builder pattern for URLs

### PR DESCRIPTION
**Pull request by allisonc07**
iOS SDK revamp!

1. Switch to Swift 5
2. Use SFWebAuthenticationSession and ASWebAuthenticationSession to provide authentication support for iOS 11+
3. Create SCUrlBuilder to create Connect URLs
4. Add new types to AuthenticationError to support more Smartcar Connect errors
5. Add jazzy documentation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,25 +37,33 @@ Then, initiate the SmartcarAuth object in the UIViewController.
...
+  redirectUri: "scafb0b7d3-807f-4c61-9b04-352e91fe3134://exchange",
   scope: ["read_vin", "read_vehicle_info", "read_odometer"],
   completion: completionHandler
 )
 let smartcar = appDelegate.smartcar
 
-// Launch Connect on the SFSafariViewController
-smartcar.launchAuthFlow(state: state, forcePrompt: false, testMode: false, viewController: viewController)
+// Generate a Connect URL
+let authUrl = authUrlBuilder().build()
```
Comment (sankethkatta) [inline README.md]: should this be?
```suggestion
let authUrl = smartcar.authUrlBuilder().build()
```

File: README.md [hunk 2]
```diff
@@ -37,25 +37,33 @@ Then, initiate the SmartcarAuth object in the UIViewController.
...
 
 ## Handling the Redirect
 
-The Connect response is returned to the app via the iOS openURL app delegate method, so you need to pipe this through to the current authorization session
+For iOS 11 and above, if you launched Connect using the _launchWebAuthSession_ method, the callback URL with the authentication code (or error) is passed back to the application by the session, and the URL is passed to the completion handler and no further action to intercept the callback is required.
+
+For iOS 10, or if you launched Connect using the _launchAuthFlow_ method, the Connect response is returned to the app via the iOS openURL app delegate method, so you need to pipe this through to the current authorization session.
 
 ```swift
 /**
```
Comment (sankethkatta) [inline README.md]: Is this code snippet still valid? (also is there a `*` missing on the second line on the doc or is that just the style of docs in swift?)

Comment (sankethkatta) [reply README.md]: Also remove the `(if any)` in the comment block below

File: SmartcarAuth/VehicleInfo.swift [hunk 3]
```diff
@@ -20,16 +20,28 @@
...
-@objc public class VehicleInfo: NSObject {
+/**
+ VehicleInfo class is used to describe the vehicle information that may be returned via query parameters in the case of authentication failure.
+ */
+@objcMembers public class VehicleInfo: NSObject {
     var vin: String?
     var make: String?
     var model: String?
-    var year: Int?
+    var year: NSNumber?
```
Comment (sankethkatta) [inline SmartcarAuth/VehicleInfo.swift]: just curious, why use `NSNumber` over `Int`?

File: SmartcarAuth/SmartcarAuth.swift [hunk 4]
```diff
@@ -20,203 +20,198 @@
...
-            queryItems.append(URLQueryItem(name: "state", value: stateString))
-        }
-
-        // if testMode specified, override self.development
-        var mode = self.development;
-        if (self.testMode != nil) {
-          // convert NSNumber to Bool
-           mode = self.testMode!.boolValue;
+    @available(iOS 11.0, *)
+    public func launchWebAuthSession(url: String, redirectUriScheme: String) {
```
Comment (sankethkatta) [inline SmartcarAuth/SmartcarAuth.swift]: get the `redirectUriScheme` from the `redirectUri`

File: SmartcarAuth/VehicleInfo.swift (hunk 3 above)
Comment (allisonc07) [reply SmartcarAuth/VehicleInfo.swift]: Int is not a supported optional type in Objective-C, so to make this class portable we need to use NSNumber!

File: README.md (hunk 2 above)
Comment (allisonc07) [reply README.md]: yes, I believe it is just the style of swift! I followed the pattern from the example linked from the jazzy docs :+1:

**Pull request merged**